src/NextGen/Framework/Receipts/DonationReceipt.php

```diff
@@ -0,0 +1,262 @@
+
+namespace Give\NextGen\Framework\Receipts;
+
+use Give\Donations\Models\Donation;
+use Give\Framework\Support\Contracts\Arrayable;
+use Give\Framework\Support\Contracts\Jsonable;
+use Give\NextGen\DonationForm\Models\DonationForm;
+use Give\NextGen\DonationForm\Repositories\DonationFormRepository;
+use Give\NextGen\Framework\Receipts\Properties\ReceiptDetail;
+use Give\NextGen\Framework\Receipts\Properties\ReceiptDetailCollection;
+
+class DonationReceipt implements Arrayable, Jsonable
+{
+    /**
+     * @var Donation
+     */
+    public $donation;
+    /**
+     * @var ReceiptDetailCollection
+     */
+    protected $donorDetails;
+    /**
+     * @var ReceiptDetailCollection
+     */
+    protected $donationDetails;
+    /**
+     * @var ReceiptDetailCollection
+     */
+    protected $additionalDetails;
+    /**
+     * @var ReceiptDetailCollection
+     */
+    protected $subscriptionDetails;
+    /**
+     * @var array
+     */
+    protected $settings;
+
+    /**
+     * @unreleased
+     */
+    public function __construct(Donation $donation)
+    {
+        $this->donation = $donation;
+        $this->donorDetails = new ReceiptDetailCollection();
+        $this->donationDetails = new ReceiptDetailCollection();
+        $this->subscriptionDetails = new ReceiptDetailCollection();
+        $this->additionalDetails = new ReceiptDetailCollection();
+        $this->settings = [];
+    }
+
+    /**
+     * @unreleased
+     */
+    public function addAdditionalDetail(string $label, $value): DonationReceipt
+    {
+        $this->additionalDetails->addDetail(new ReceiptDetail($label, $value));
+
+        return $this;
+    }
+
+    /**
+     * @unreleased
+     *
+     * @param  string  $key
+     * @param  mixed  $value
+     *
+     * @return DonationReceipt
+     */
+    public function addSetting(string $key, $value): DonationReceipt
+    {
+        $this->settings[$key] = $value;
+
+        return $this;
+    }
+
+    /**
+     * @unreleased
+     */
+    public function toArray(): array
+    {
+        return [
+            'settings' => $this->settings,
+            'donorDetails' => $this->donorDetails->toArray(),
+            'donationDetails' => $this->donationDetails->toArray(),
+            'subscriptionDetails' => $this->subscriptionDetails->toArray(),
+            'additionalDetails' => $this->additionalDetails->toArray(),
+        ];
+    }
+
+    /**
+     * @unreleased
+     */
+    public function toJson($options = 0): string
+    {
+        return json_encode($this->toArray(), $options);
+    }
+
+    /**
+     * @unreleased
+     */
+    protected function getCustomFields(): array
+    {
+        if (give(DonationFormRepository::class)->isLegacyForm($this->donation->formId)) {
+            return [];
+        }
+
+        /** @var DonationForm $form */
+        $form = DonationForm::find($this->donation->formId);
+
+        $customFields = [];
+
+        $form->schema()->walkFields(function ($field) use (&$customFields) {
```

Comment:
Would it make sense to `filter` and `map` `$form->schema()->getFields()` here, as opposed to walking the fields and building the `$customFields` array?

The `walkFields` method is more so intended to make changes to the Fields passed by reference.

Reply:
Do we really not have a general use `Collection` class with `filter` and `map` methods in Give core? Hm...

Reply:
resolved 6bedb72c